File: src/movegen/pawn.rs
use crate::{
    bitboard::{clear_bit, get_bit, set_bit},
    constants::{A1, A2, A7, A8, H1, H2, H7, H8, NOT_FILE_A, NOT_FILE_H},
    enums::{Color, PieceAndColor},
    position::Position,
};

use super::Move;

pub struct PawnMoveGen {
    attack_table: [[u64; 64]; 2],
}

impl PawnMoveGen {
    fn mask_attacks(color: Color, square: u8) -> u64 {
        let mut attacks = 0u64;
        let mut bitboard = 0u64;

        set_bit(&mut bitboard, square);

        if color == Color::White {
            attacks |= (bitboard >> 7) & NOT_FILE_A;
            attacks |= (bitboard >> 9) & NOT_FILE_H;
        } else {
            attacks |= (bitboard << 7) & NOT_FILE_H;
            attacks |= (bitboard << 9) & NOT_FILE_A;
        }

        attacks
    }

    fn generate_attack_table() -> [[u64; 64]; 2] {
        let mut table = [[0; 64]; 2];

        for square in 0..64u8 {
            table[Color::White as usize][square as usize] =
                Self::mask_attacks(Color::White, square);
            table[Color::Black as usize][square as usize] =
                Self::mask_attacks(Color::Black, square);
        }

        table
    }

    pub fn attack_table(&self) -> &[[u64; 64]; 2] {
        &self.attack_table
    }

    pub fn generate_moves(&self, position: &Position) -> Vec<Move> {
        let color = position.color_to_move();

        let piece = match color {
            Color::White => PieceAndColor::WhitePawn,
            Color::Black => PieceAndColor::BlackPawn,
            _ => unreachable!(),
        };

        let mut moves = Vec::new();

        let mut bitboard = position.bitboards().piece(piece);

        while bitboard != 0 {
            let source_square = bitboard.trailing_zeros() as u8;

            let promotion_rank = match color {
                Color::White => A8..=H8,
                Color::Black => A1..=H1,
                _ => unreachable!(),
            };

            let home_rank = match color {
                Color::White => A2..=H2,
                Color::Black => A7..=H7,
                _ => unreachable!(),
            };

            if promotion_rank.contains(&source_square) {
                continue;
            }

            let target_square = match color {
                Color::White => source_square - 8,
                Color::Black => source_square + 8,
                _ => unreachable!(),
            };

            if !get_bit(position.bitboards().color(Color::Both), target_square) {
                if promotion_rank.contains(&target_square) {
                    moves.push(Move::new(
                        source_square,
                        target_square,
                        piece,
                        Some(PieceAndColor::WhiteQueen),
                        false,
                        false,
                        false,
                        false,
                    ));

                    moves.push(Move::new(
                        source_square,
                        target_square,
                        piece,
                        Some(PieceAndColor::WhiteRook),
                        false,
                        false,
                        false,
                        false,
                    ));

                    moves.push(Move::new(
                        source_square,
                        target_square,
                        piece,
                        Some(PieceAndColor::WhiteBishop),
                        false,
                        false,
                        false,
                        false,
                    ));

                    moves.push(Move::new(
                        source_square,
                        target_square,
                        piece,
                        Some(PieceAndColor::WhiteKnight),
                        false,
                        false,
                        false,
                        false,
                    ));
                } else {
                    moves.push(Move::new(
                        source_square,
                        target_square,
                        piece,
                        None,
                        false,
                        false,
                        false,
                        false,
                    ));

                    if home_rank.contains(&source_square) {
                        let double_target_square = match color {
                            Color::White => source_square - 16,
                            Color::Black => source_square + 16,
                            _ => unreachable!(),
                        };

                        if !get_bit(
                            position.bitboards().color(Color::Both),
                            double_target_square,
                        ) {
                            moves.push(Move::new(
                                source_square,
                                double_target_square,
                                piece,
                                None,
                                false,
                                false,
                                false,
                                true,
                            ));
                        }
                    }
                }
            }

            let attacks_mask = self.attack_table[color as usize][source_square as usize];

            let mut attacks = attacks_mask & position.bitboards().color(!color);

            while attacks != 0 {
                let capture_square = attacks.trailing_zeros() as u8;

                if promotion_rank.contains(&capture_square) {
                    moves.push(Move::new(
                        source_square,
                        capture_square,
                        piece,
                        Some(PieceAndColor::WhiteQueen),
                        true,
                        false,
                        false,
                        false,
                    ));

                    moves.push(Move::new(
                        source_square,
                        capture_square,
                        piece,
                        Some(PieceAndColor::WhiteRook),
                        true,
                        false,
                        false,
                        false,
                    ));

                    moves.push(Move::new(
                        source_square,
                        capture_square,
                        piece,
                        Some(PieceAndColor::WhiteBishop),
                        true,
                        false,
                        false,
                        false,
                    ));

                    moves.push(Move::new(
                        source_square,
                        capture_square,
                        piece,
                        Some(PieceAndColor::WhiteKnight),
                        true,
                        false,
                        false,
                        false,
                    ));
                } else {
                    moves.push(Move::new(
                        source_square,
                        capture_square,
                        piece,
                        None,
                        true,
                        false,
                        false,
                        false,
                    ));
                }

                clear_bit(&mut attacks, capture_square);
            }

            if let Some(en_passant_square) = position.en_passant_square() {
                if get_bit(attacks_mask, en_passant_square) {
                    moves.push(Move::new(
                        source_square,
                        en_passant_square,
                        piece,
                        None,
                        true,
                        true,
                        false,
                        false,
                    ));
                }
            }

            clear_bit(&mut bitboard, source_square);
        }

        moves
    }

    pub fn new() -> Self {
        Self {
            attack_table: Self::generate_attack_table(),
        }
    }
}

File: internal/movegen/pawn.go
package movegen

import (
	"github.com/notmalte/psce/internal/bitboard"
	"github.com/notmalte/psce/internal/constants"
	"github.com/notmalte/psce/internal/position"
	"math/bits"
)

type PawnMoveGen struct {
	attackTable [2][64]uint64
}

func (pmg *PawnMoveGen) generateAttackTable() [2][64]uint64 {
	table := [2][64]uint64{}

	for square := range uint8(64) {
		table[constants.ColorWhite][square] = pmg.maskAttacks(constants.ColorWhite, square)
		table[constants.ColorBlack][square] = pmg.maskAttacks(constants.ColorBlack, square)
	}

	return table
}

func (pmg *PawnMoveGen) maskAttacks(color uint8, square uint8) uint64 {
	attacks := uint64(0)
	bb := uint64(0)
	bitboard.SetBit(&bb, square)

	if color == constants.ColorWhite {
		attacks |= (bb >> 7) & constants.NotFileA
		attacks |= (bb >> 9) & constants.NotFileH
	} else {
		attacks |= (bb << 7) & constants.NotFileH
		attacks |= (bb << 9) & constants.NotFileA
	}

	return attacks
}

func (pmg *PawnMoveGen) GetAttacks(square uint8, color uint8) uint64 {
	return pmg.attackTable[color][square]
}

func NewPawnMoveGen() *PawnMoveGen {
	pmg := &PawnMoveGen{}
	pmg.attackTable = pmg.generateAttackTable()

	return pmg
}

func (pmg *PawnMoveGen) GeneratePseudoLegalMoves(pos *position.Position) []Move {
	isWhite := pos.ColorToMove == constants.ColorWhite

	var otherColor uint8
	var piece uint8
	var promotionRowStart uint8
	var promotionRowEnd uint8
	var homeRowStart uint8
	var homeRowEnd uint8
	if isWhite {
		otherColor = constants.ColorBlack
		piece = constants.WhitePawn
		promotionRowStart = constants.A8
		promotionRowEnd = constants.H8
		homeRowStart = constants.A2
		homeRowEnd = constants.H2
	} else {
		otherColor = constants.ColorWhite
		piece = constants.BlackPawn
		promotionRowStart = constants.A1
		promotionRowEnd = constants.H1
		homeRowStart = constants.A7
		homeRowEnd = constants.H7
	}

	moves := []Move{}
	bb := pos.PieceBitboards[piece]

	/*
		TODO: this can be done by shifting and masking,
			<< +-8 for single push, << +-16 for double push,
			& with constants.NotFileA and constants.NotFileH for file checks
		 	<< +-7 and << +-9 for captures
	*/

	for bb != 0 {
		fromSquare := uint8(bits.TrailingZeros64(bb))

		if fromSquare >= promotionRowStart && fromSquare <= promotionRowEnd {
			continue
		}

		var toSquare uint8
		if isWhite {
			toSquare = fromSquare - 8
		} else {
			toSquare = fromSquare + 8
		}

		if !bitboard.GetBit(pos.ColorBitboards[constants.ColorBoth], toSquare) {
			if toSquare >= promotionRowStart && toSquare <= promotionRowEnd {
				var promotionPieces [4]uint8
				if isWhite {
					promotionPieces = [4]uint8{constants.WhiteQueen, constants.WhiteRook, constants.WhiteBishop, constants.WhiteKnight}
				} else {
					promotionPieces = [4]uint8{constants.BlackQueen, constants.BlackRook, constants.BlackBishop, constants.BlackKnight}
				}

				for _, promotionPiece := range promotionPieces {
					moves = append(moves, Move{
						FromSquare:     fromSquare,
						ToSquare:       toSquare,
						Piece:          piece,
						Flags:          FlagPromotion,
						PromotionPiece: promotionPiece,
					})
				}
			} else {
				moves = append(moves, Move{
					FromSquare: fromSquare,
					ToSquare:   toSquare,
					Piece:      piece,
				})

				if fromSquare >= homeRowStart && fromSquare <= homeRowEnd {
					var doublePushSquare uint8
					if isWhite {
						doublePushSquare = fromSquare - 16
					} else {
						doublePushSquare = fromSquare + 16
					}

					if !bitboard.GetBit(pos.ColorBitboards[constants.ColorBoth], doublePushSquare) {
						moves = append(moves, Move{
							FromSquare: fromSquare,
							ToSquare:   doublePushSquare,
							Piece:      piece,
							Flags:      FlagDoublePawnPush,
						})
					}
				}
			}
		}

		attacks := pmg.GetAttacks(fromSquare, pos.ColorToMove)
		captures := attacks & pos.ColorBitboards[otherColor]

		for captures != 0 {
			toSquare = uint8(bits.TrailingZeros64(captures))

			if toSquare >= promotionRowStart && toSquare <= promotionRowEnd {
				var promotionPieces [4]uint8
				if isWhite {
					promotionPieces = [4]uint8{constants.WhiteQueen, constants.WhiteRook, constants.WhiteBishop, constants.WhiteKnight}
				} else {
					promotionPieces = [4]uint8{constants.BlackQueen, constants.BlackRook, constants.BlackBishop, constants.BlackKnight}
				}

				for _, promotionPiece := range promotionPieces {
					moves = append(moves, Move{
						FromSquare:     fromSquare,
						ToSquare:       toSquare,
						Piece:          piece,
						Flags:          FlagPromotion | FlagCapture,
						PromotionPiece: promotionPiece,
					})
				}
			} else {
				moves = append(moves, Move{
					FromSquare: fromSquare,
					ToSquare:   toSquare,
					Piece:      piece,
					Flags:      FlagCapture,
				})
			}

			bitboard.ClearBit(&captures, toSquare)
		}

		if pos.EnPassantSquare != constants.NoSquare {
			if bitboard.GetBit(attacks, pos.EnPassantSquare) {
				moves = append(moves, Move{
					FromSquare: fromSquare,
					ToSquare:   pos.EnPassantSquare,
					Piece:      piece,
					Flags:      FlagEnPassant | FlagCapture,
				})
			}
		}

		bitboard.ClearBit(&bb, fromSquare)
	}

	return moves
}
